// @swup/forms-plugin/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _plugin = require('@swup/plugin');

var _plugin2 = _interopRequireDefault(_plugin);

var _delegate = require('delegate');

var _delegate2 = _interopRequireDefault(_delegate);

var _utils = require('swup/lib/utils');

var _Link = require('swup/lib/helpers/Link');

var _Link2 = _interopRequireDefault(_Link);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var FormPlugin = function (_Plugin) {
    _inherits(FormPlugin, _Plugin);

    function FormPlugin(options) {
        _classCallCheck(this, FormPlugin);

        var _this = _possibleConstructorReturn(this, (FormPlugin.__proto__ || Object.getPrototypeOf(FormPlugin)).call(this));

        _this.name = "FormsPlugin";


        var defaultOptions = {
            formSelector: 'form[data-swup-form]'
        };

        _this.options = _extends({}, defaultOptions, options);
        return _this;
    }

    _createClass(FormPlugin, [{
        key: 'mount',
        value: function mount() {
            var swup = this.swup;

            // add empty handlers array for submitForm event
            swup._handlers.submitForm = [];
            swup._handlers.openFormSubmitInNewTab = [];

            // register handler
            swup.delegatedListeners.formSubmit = (0, _delegate2.default)(document, this.options.formSelector, 'submit', this.onFormSubmit.bind(this));
        }
    }, {
        key: 'unmount',
        value: function unmount() {
            swup.delegatedListeners.formSubmit.destroy();
        }
    }, {
        key: 'onFormSubmit',
        value: function onFormSubmit(event) {
            var swup = this.swup;

            // no control key pressed
            if (!event.metaKey) {
                var form = event.target;
                var formData = new FormData(form);
                var actionAttribute = form.getAttribute('action') || window.location.href;
                var methodAttribute = form.getAttribute('method') || 'GET';
                var link = new _Link2.default(actionAttribute);

                // fomr
                swup.triggerEvent('submitForm', event);

                event.preventDefault();

                if (link.getHash() != '') {
                    swup.scrollToElement = link.getHash();
                }

                // get custom transition from data
                var customTransition = form.getAttribute('data-swup-transition');

                if (methodAttribute.toLowerCase() != 'get') {
                    // remove page from cache
                    swup.cache.remove(link.getAddress());

                    // send data
                    swup.loadPage({
                        url: link.getAddress(),
                        method: methodAttribute,
                        data: formData,
                        customTransition: customTransition
                    });
                } else {
                    // create base url
                    var url = link.getAddress() || window.location.href;
                    var inputs = (0, _utils.queryAll)('input, select', form);
                    if (url.indexOf('?') == -1) {
                        url += '?';
                    } else {
                        url += '&';
                    }

                    // add form data to url
                    inputs.forEach(function (input) {
                        if (input.type == 'checkbox' || input.type == 'radio') {
                            if (input.checked) {
                                url += encodeURIComponent(input.name) + '=' + encodeURIComponent(input.value) + '&';
                            }
                        } else {
                            url += encodeURIComponent(input.name) + '=' + encodeURIComponent(input.value) + '&';
                        }
                    });

                    // remove last "&"
                    url = url.slice(0, -1);

                    // remove page from cache
                    swup.cache.remove(url);

                    // send data
                    swup.loadPage({
                        url: url,
                        customTransition: customTransition
                    });
                }
            } else {
                swup.triggerEvent('openFormSubmitInNewTab', event);
            }
        }
    }]);

    return FormPlugin;
}(_plugin2.default);

exports.default = FormPlugin;
